### PR TITLE
linux-v4l2: Fix fallback framerate for camera

### PR DESCRIPTION
**Pull request by 9p4**
### Description

Fixes very large log sizes and STDOUT spam when camera does not report framerate.

Code taken from https://github.com/Mikael-Lovqvist/obs-studio/commit/3cd5742be5168f625558c14c1b38fedf23116277

This adds a fallback framerate if the camera does not report one. It defaults to 10 frames per second, but can be manually changed later in the UI (included).

### Motivation and Context

- Fixes https://github.com/obsproject/obs-studio/issues/4926
- Fixes https://github.com/obsproject/obs-studio/issues/8397

### How Has This Been Tested?

Tested on `Linux hostname 5.16.11-arch1-1 #1 SMP PREEMPT Thu, 24 Feb 2022 02:18:20 +0000 x86_64 GNU/Linux`. Built with `mkdir build && cd build && cmake -DCMAKE_INSTALL_PREFIX="/usr" -DBUILD_BROWSER=OFF -DBUILD_VST=OFF -DDISABLE_VLC=ON .. && make -j$(nproc) && mkdir dist && make install DESTDIR=dist`. Tested by checking log sizes and STDOUT for affected camera. Log was as expected: without the hundreds of thousands of error messages.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by kkartaltepe** [comment]
Can you please provide more information about the device where this occurs? What framerates does it support that is resulting in us picking 0/0?

**Comment by 9p4** [comment]
The camera seems to not report any supported framerate whatsoever, which is causing OBS to default to a framerate of zero. The culprit in question is (I think, I'm not quite sure) a Logitech QuickCam Express (320x240 @ 30 frame/s,640x480 @ 15 frame/s).

**Comment by 9p4** [comment]
Converting to draft since it seems as if it's not particularly reliable at the moment.

**Comment by 9p4** [comment]
> What tools have you used to determine the reported framerates? Why does this work in any other software if it reports no supported framerates?

I have not determined the reported framerates. I took those values from the datasheet online. I chose 10 since it seemed like a reasonable default. Keep in mind that this is a stopgap to prevent filling up the logs with a strange webcam, and the user is meant to change the FPS of the camera to the actual value once the application is loaded.

**Comment by jp9000** [comment]
I have to be honest, I care more about fixing the log file issue than I do trying to support an ancient camera

File: plugins/linux-v4l2/v4l2-input.c [hunk 1]
```diff
@@ -39,6 +39,8 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "v4l2-helpers.h"
 #include "v4l2-mjpeg.h"
 
+#define FALLBACK_FRAMERATE 10
```
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-input.c:42]: Use 30 to match chromes implementation

File: plugins/linux-v4l2/v4l2-input.c [hunk 2]
```diff
@@ -169,7 +171,7 @@ static void *v4l2_thread(void *vptr)
 	struct v4l2_buffer buf;
 	struct obs_source_frame out;
 	size_t plane_offsets[MAX_AV_PLANES];
-	int fps_num, fps_denom;
+	uint fps_num, fps_denom;
```
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-input.c:174]: Why did you change the type?

File: plugins/linux-v4l2/v4l2-input.c [hunk 3]
```diff
@@ -1008,11 +1010,18 @@ static void v4l2_init(struct v4l2_data *data)
...
+		     FALLBACK_FRAMERATE);
+		data->framerate = v4l2_pack_tuple(1, FALLBACK_FRAMERATE);
+		if (v4l2_set_framerate(data->dev, &data->framerate) < 0) {
+			blog(LOG_ERROR, "Unable to set framerate");
+			goto fail;
+		}
 	}
 	v4l2_unpack_tuple(&fps_num, &fps_denom, data->framerate);
+
+	blog(LOG_INFO, "fps_num: %d", fps_num);
```
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-input.c:1024]: Remove this added logging.

File: plugins/linux-v4l2/v4l2-input.c [hunk 4]
```diff
@@ -1008,11 +1010,18 @@ static void v4l2_init(struct v4l2_data *data)
...
 
 	/* set framerate */
-	if (v4l2_set_framerate(data->dev, &data->framerate) < 0) {
-		blog(LOG_ERROR, "Unable to set framerate");
-		goto fail;
+	if (data->framerate == 0) {
+		blog(LOG_ERROR, "Framerate is not set, falling back to %i",
+		     FALLBACK_FRAMERATE);
+		data->framerate = v4l2_pack_tuple(1, FALLBACK_FRAMERATE);
+		if (v4l2_set_framerate(data->dev, &data->framerate) < 0) {
```
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-input.c:1017]: This should be outside of the if statement to preserve functionality for non-broken webcams.

**Review by kkartaltepe — COMMENTED** [review]
I looked at chrome and it seems they do a similar fallback for broken cameras so I guess there is no reason not to do this.

**Comment by 9p4** [comment]
I'll test out this commit with the old webcam when I get back home for spring break.

File: plugins/linux-v4l2/v4l2-input.c [hunk 5]
```diff
@@ -1025,6 +1027,11 @@ static void v4l2_init(struct v4l2_data *data)
 	blog(LOG_INFO, "Linesize: %d Bytes", data->linesize);
 
 	/* set framerate */
+	if (data->framerate == 0) {
```
Comment (kkartaltepe) [inline plugins/linux-v4l2/v4l2-input.c:1030]: I think it got mixed when you were applying my suggestions. We should move this after the v4l2_set_framerate (tries to apply the chosen fps and returns the resulting fps, which may be 0) and before the unpack on L1039.

Comment (9p4) [reply plugins/linux-v4l2/v4l2-input.c:1030]: whoops. should be fixed now

**Comment by 9p4** [comment]
Can confirm that this works.

**Review by kkartaltepe — APPROVED** [review]

**Comment by kkartaltepe** [comment]
Please rebase instead of including merge commits in your PR.

**Comment by 9p4** [comment]
Done.
